File: apps/api/src/lib/object-id.ts
import { ObjectId } from "mongodb";
import { BadRequestError } from "./errors";

/**
 * Ép id chuỗi → ObjectId. id sai định dạng → BadRequestError (400) thay vì để
 * `new ObjectId(...)` ném BSONError rồi rơi vào nhánh 500 của error handler.
 */
export function toObjectId(id: string): ObjectId {
  if (!ObjectId.isValid(id)) {
    throw new BadRequestError(`id không hợp lệ: ${id}`);
  }
  return new ObjectId(id);
}

File: services/agent-go/internal/mongo/objectid.go
package mongo

import (
	"fmt"

	"go.mongodb.org/mongo-driver/v2/bson"
)

// ToObjectID ép hex 24 ký tự → bson.ObjectID. id sai định dạng → error rõ ràng
// (caller/gateway map thành 400 thay vì 500).
func ToObjectID(id string) (bson.ObjectID, error) {
	oid, err := bson.ObjectIDFromHex(id)
	if err != nil {
		return bson.NilObjectID, fmt.Errorf("id không hợp lệ: %s", id)
	}
	return oid, nil
}
